float64[] step_size
uint64[] occupied_indexes # index0, index1, ...